src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -1,234 +1,371 @@
-    *
-    * @return The head of the deque.
-    *
-    * @throws IllegalStateException if the thread holding the write lock
-    *         attempts to call this method.
-    * @throws InterruptedException
-    */
     @Override
-    public E borrowItem() throws InterruptedException, IllegalStateException {
-        if (lock.isWriteLockedByCurrentThread()) {
-          throw new IllegalStateException("The current implementation has a risk of deadlock if you attempt to borrow a JRuby instance while holding the write lock!");
+    public E borrowItem() throws InterruptedException {
+        E item = null;
+        final ReentrantLock lock = this.lock;
+        lock.lock();
+        try {
+            final Thread currentThread = Thread.currentThread();
+            do {
+                if (isWriteLockHeldByAnotherThread(currentThread)) {
+                    notWriteLocked.await();
+                } else if (liveQueue.size() < 1) {
+                    notEmpty.await();
+                } else {
+                    item = liveQueue.removeFirst();
+                }
+            } while (item == null);
+        } finally {
+            lock.unlock();
         }
-        E item = liveQueue.takeFirst();
-        lock.readLock().lock();
+
         return item;
     }
 
-    /**
-    * This method is analagous to <tt>pollFirst</tt> in the
-    * <tt>LinkedBlockingDeque</tt> class, but also causes a read lock to be
-    * locked from the <tt>ReentrantReadWriteLock</tt>.
-    *
-    * @param timeout how long to wait before giving up, in units of unit
-    * @param unit a <tt>TimeUnit</tt> determining how to interpret the
-    *        <tt>timeout parameter</tt>
-    *
-    * @return The head of the deque, or <tt>null</tt> if the specified waiting
-    *         time elapses before an element is available.
-    *
-    * @throws IllegalStateException if the thread holding the write lock
-    *         attempts to call this method.
-    * @throws InterruptedException
-    */
     @Override
-    public E borrowItemWithTimeout(long timeout, TimeUnit unit) throws InterruptedException, IllegalStateException {
-        if (lock.isWriteLockedByCurrentThread()) {
-          throw new IllegalStateException("The current implementation has a risk of deadlock if you attempt to borrow a JRuby instance while holding the write lock!");
+    public E borrowItemWithTimeout(long timeout, TimeUnit unit) throws
+            InterruptedException {
+        E item = null;
+        final ReentrantLock lock = this.lock;
+        long remainingMaxTimeToWait = unit.toNanos(timeout);
+
+        // `lock.lockInterruptibly()` is called here as opposed to just
+        // `lock.lock` to follow the pattern that the JDK's
+        // `LinkedBlockingDeque` does for a timed poll from a deque.  See:
+        // http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/concurrent/LinkedBlockingDeque.java#l516
+        lock.lockInterruptibly();
```

Comment:
Is it weird that `lockInterruptibly` isn't used elsewhere, especially `borrowItem`?